components/org.wso2.carbon.identity.oauth.endpoint/src/test/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpointTest.java

```diff
@@ -221,36 +328,238 @@ public Object answer(InvocationOnMock invocation) throws Throwable {
+                        HttpServletResponse.SC_FOUND, null},
+
+                {"approve", "http://localhost:8080/redirect", new HashSet<>(Arrays.asList("scope1")),
+                        HttpServletResponse.SC_FOUND, null},
+
+                {"approve", "{\"url\":\"http://localhost:8080/redirect\"}", new HashSet<>(Arrays.asList("openid")),
+                        HttpServletResponse.SC_OK, null},
+
+                {"approve", "{\"url\":\"http://localhost:8080/redirect\"}", new HashSet<>(Arrays.asList("scope1")),
+                        HttpServletResponse.SC_OK, null},
+        };
+    }
 
+    @Test(dataProvider = "provideConsentData")
+    public void testUserConsentResponse(String consent, String redirectUrl, Set<String> scopes,
+                                        int expectedStatus, String expectedError) throws Exception {
+        mockStatic(SessionDataCache.class);
+        when(SessionDataCache.getInstance()).thenReturn(sessionDataCache);
+        SessionDataCacheKey consentDataCacheKey = new SessionDataCacheKey(this.sessionDataKeyConsent);
+        when(sessionDataCache.getValueFromCache(consentDataCacheKey)).thenReturn(consentCacheEntry);
+
+        Map<String, String[]> requestParams = new HashedMap();
+        Map<String, Object> requestAttributes = new HashedMap();
+
+        requestParams.put(OAuthConstants.SESSION_DATA_KEY_CONSENT, new String[]{sessionDataKeyConsent});
+        requestParams.put(FrameworkConstants.RequestParams.TO_COMMONAUTH, new String[]{"false"});
+        requestParams.put(OAuthConstants.OAuth20Params.SCOPE, new String[]{"openid"});
+        requestParams.put(OAuthConstants.Prompt.CONSENT, new String[]{consent});
+
+        requestAttributes.put(FrameworkConstants.RequestParams.FLOW_STATUS, AuthenticatorFlowStatus.INCOMPLETE);
+
+        mockHttpRequest(requestParams, requestAttributes);
+
+        OAuth2Parameters oAuth2Params = setOAuth2Parameters(scopes, appName, RESPONSE_MODE_FORM_POST, redirectUrl);
+
+        when(consentCacheEntry.getoAuth2Parameters()).thenReturn(oAuth2Params);
+        when(consentCacheEntry.getLoggedInUser()).thenReturn(new AuthenticatedUser());
+
+        mockStatic(OpenIDConnectUserRPStore.class);
+        when(OpenIDConnectUserRPStore.getInstance()).thenReturn(openIDConnectUserRPStore);
+        doNothing().when(openIDConnectUserRPStore).putUserRPToStore(any(AuthenticatedUser.class),
+                anyString(), anyBoolean(), anyString());
+
+        mockStatic(OAuthServerConfiguration.class);
+        when(OAuthServerConfiguration.getInstance()).thenReturn(oAuthServerConfiguration);
+        when(oAuthServerConfiguration.getPersistenceProcessor()).thenReturn(tokenPersistenceProcessor);
+        when(tokenPersistenceProcessor.getProcessedClientId(anyString())).thenAnswer(new Answer<Object>(){
+            @Override
+            public Object answer(InvocationOnMock invocation) throws Throwable {
```

Comment:
Then do call real ?

Reply:
Need one more level of mocking to call the real method since tokenPersistenceProcessor is mocking an interface.